packages/framework-provider-aws/src/library/events-adapter.ts

```diff
@@ -101,28 +103,78 @@ export async function storeEvents(
   logger.debug('EventEnvelopes stored')
 }
 
+export async function storeSnapshot(
+  dynamoDB: DynamoDB.DocumentClient,
+  snapshotEnvelope: EntitySnapshotEnvelope,
+  config: BoosterConfig
+): Promise<void> {
+  try {
+    const logger = getLogger(config, 'EventsAdapter#storeSnapshot')
+    logger.debug('Storing the following snapshot:', snapshotEnvelope)
+
+    const partitionKey = partitionKeyForEntitySnapshot(snapshotEnvelope.entityTypeName, snapshotEnvelope.entityID)
+    /**
+     * The sort key of the snapshot matches the sort key of the last event that generated it.
+     * Entity snapshots can be potentially created by competing processes, so we need to make sure
+     * that no matter what snapshot we find, we can always rebuild the correct state by
+     * replaying all the events that happened after the snapshot we take as the origin.
+     * This way of storing the data makes snapshot caching an idempotent operation, and allows us to
+     * aggressively caching snapshots without worrying about insertion order.
+     */
+    const sortKey = snapshotEnvelope.snapshottedEventPersistedAt
+    await dynamoDB
+      .put({
+        TableName: config.resourceNames.eventsStore,
+        ConditionExpression: `${eventsStoreAttributes.partitionKey} <> :partitionKey AND ${eventsStoreAttributes.sortKey} <> :sortKey`,
+        ExpressionAttributeValues: {
+          ':partitionKey': partitionKey,
+          ':sortKey': sortKey,
+        },
+        Item: {
+          ...snapshotEnvelope,
+          [eventsStoreAttributes.partitionKey]: partitionKey,
+          [eventsStoreAttributes.sortKey]: sortKey,
```

Comment:
One detail here: While we're introducing a new `snapshottedEventPersistedAt` field and using it as the sort key value, the `eventsStoreAttributes.sortKey` constant is still `createdAt`. This means that at the time of writing, the value stored will override whatever value we set in `createdAt` and replace It with the value set in `snapshottedEventPersistedAt`. I think that this is important to keep backward compatibility, but I'm totally open to other thoughts on this.

Reply:
Not sure if we will need it or not, but we are not going to persist a timestamp with the snapshot creation, are we?

Reply:
I rolled back the introduction of the `persistedAt` field to use the `createdAt` field again, but I'm still updating the `createdAt` to match the one from the entity. As we're now generating snapshots from multiple threads, this is the only way to know if a specific snapshot has been generated before, but it could make sense to also keep an extra timestamp with the snapshot persist time. I'll add that field here.